Add implementation of ConvexMethod members

hungarianhomie opened the pull request.

hungarianhomie committed:
Implementation of the below methods of ConvexMethod

Note: unable to test for some reason, even on clean branch

get_detections_vector()
get_convex_hull()
determine_scenario()

hungarianhomie committed:
Complete implementation of rest of ConvexMethod classes

Code builds but is NOT tested yet!

Added Todo in obj planner node to sort the path by x distance

hungarianhomie commented:
This is still not unit tested!
Some test will be added in follow on commits

hungarianhomie commented on src/ConvexMethod.cpp:
```diff
@@ -22,47 +52,204 @@ std::optional<LeftRightResults> ConvexMethod::classify(const geometry_msgs::msg:
 
-Scenario ConvexMethod::determine_scenario() {
+Scenario ConvexMethod::determine_scenario(const std::vector<cv::Point2d>& detections_2d) {
     //TODO impl
```
Remove TODO and others missed

hungarianhomie commented on include/obj_planner/ConvexMethod.hpp:
```diff
@@ -15,53 +22,93 @@ enum class Scenario : std::uint8_t { kStraight = 0, kLeft = 1U, kRight = 2U };
+    virtual std::vector<cv::Point2d>& get_inside_detections_result(LeftRightResults& classification) = 0;
+
+private:
+    std::size_t find_end_segment_index(const std::size_t start_idx, const std::vector<cv::Point2d>& convex_hull);
```
Add some description here and other methods

andyblarblar commented on src/ConvexMethod.cpp:
```diff
@@ -1,18 +1,48 @@
+
+/// Only let number reside in [0, 1]
+void clamp_between_0_and_1(double& number) {
+    if (number < 0.0) {
```
Not a huge deal, but std::clamp does exist

hungarianhomie replied:
Completely forgot about std::clamp, Ill update to use that instead

hungarianhomie replied:
For now I didn't update this to prioritize getting it tested and working.

andyblarblar commented on src/ConvexMethod.cpp:
```diff
@@ -22,47 +52,204 @@ std::optional<LeftRightResults> ConvexMethod::classify(const geometry_msgs::msg:
-    return {};
+LeftRightResults StraightScenarioClassifier::classify(const std::vector<cv::Point2d>& convex_hull,
+                                                      const std::vector<cv::Point2d>& detections_2d) {
+    (void)convex_hull;  // Unused but needed to satisfy interface. This hurts me deeply
```
Whats the reasoning for this? Never seen this before.

hungarianhomie replied:
Thanks for making this better!

andyblarblar committed:
Replace void cast with maybe unused

hungarianhomie committed:
Add unit tests to convex method

A few updates after finding some bugs in unit tests.
The parameterized tests test the performance of the algo.
1 of them is expected to have some misclassification, this
can be (hopefully) remedied later with parameter tuning.
